### PR TITLE
feat(skills): sync skill directories for Claude, Cursor, Codex + skill remove verb

### PR DESCRIPTION
**Pull request by chrisleekr**
## Summary

- Adds end-to-end round-trip for user-created skill directories under `~/.claude/skills/`, `~/.codex/skills/`, and `~/.cursor/skills/` alongside the already-supported Copilot skills, and introduces a new `agentsync skill remove <agent> <name>` CLI verb for explicit vault removal.
- Extracts a shared five-gate `skills-walker.ts` module (dot-skip → root-symlink rejection → SKILL.md sentinel via lstat → never-sync interior scan → symlink-filtered tar) and retrofits Copilot onto it so every skill-bearing agent inherits identical safety rules. `src/core/tar.ts::archiveDirectory` gains an opt-in `skipSymlinks` flag; `src/commands/push.ts`' Phase-1 gate escalates walker-level `never-sync inside skill:` warnings to fatal aborts; `src/commands/doctor.ts` now reports readability of the three new skills directories.
- Push is **additive-by-default** (FR-011) — a local delete never mutates the vault. Pull is **extract-only** (FR-013) — even when a vault artifact is gone, `applyXxxVault` leaves existing local skill directories untouched. The `skill remove` verb is the only operation that takes a skill out of the vault, and it never touches any local skill directory on any machine (FR-012).

## Spec artifacts

- [spec.md](../blob/20260411-002222-agent-skills-sync/specs/20260411-002222-agent-skills-sync/spec.md)
- [plan.md](../blob/20260411-002222-agent-skills-sync/specs/20260411-002222-agent-skills-sync/plan.md)
- [research.md](../blob/20260411-002222-agent-skills-sync/specs/20260411-002222-agent-skills-sync/research.md)
- [data-model.md](../blob/20260411-002222-agent-skills-sync/specs/20260411-002222-agent-skills-sync/data-model.md)
- [contracts/walker-interface.md](../blob/20260411-002222-agent-skills-sync/specs/20260411-002222-agent-skills-sync/contracts/walker-interface.md)
- [contracts/skill-remove-cli.md](../blob/20260411-002222-agent-skills-sync/specs/20260411-002222-agent-skills-sync/contracts/skill-remove-cli.md)
- [contracts/vault-paths.md](../blob/20260411-002222-agent-skills-sync/specs/20260411-002222-agent-skills-sync/contracts/vault-paths.md)
- [quickstart.md](../blob/20260411-002222-agent-skills-sync/specs/20260411-002222-agent-skills-sync/quickstart.md)
- [tasks.md](../blob/20260411-002222-agent-skills-sync/specs/20260411-002222-agent-skills-sync/tasks.md)

## Requirements coverage

Every FR/SC listed below has at least one automated regression test in this PR. The full matrix is in `tasks.md` at the bottom of the file.

| Ref | Guarantee | Proof |
|-----|-----------|-------|
| FR-001/003/004 | Round-trip for Claude, Codex, Cursor | `claude.test.ts`, `codex.test.ts`, `cursor.test.ts` happy-path snapshots |
| FR-002 | Real `SKILL.md` sentinel required | `skills-walker.test.ts` rows 4–6, agent-layer mirror tests |
| FR-006 | Never-sync inside a skill → fatal push abort | `push.test.ts` T008 |
| FR-007 / SC-003 | Status drift for skill artifacts | `status.test.ts` T035 |
| FR-008 | `doctor` checks skills directories | `doctor.test.ts` T012 |
| FR-009 | Missing skills dir is a no-op | T014(5) / T017(5) / T020(5) per agent |
| FR-010 | Cursor `~/.cursor/skills-cursor/` never touched | `cursor.test.ts` T021 |
| FR-011 / SC-006 | Additive-default: local delete ≠ vault delete | `push.test.ts` T036 |
| FR-012 | `skill remove` never touches local skills | `skill.test.ts` "leave-local-alone" |
| FR-013 | Pull is extract-only | `integration.test.ts` T026 FR-013 test |
| FR-016 outer | Symlinked skill root silently skipped | walker T006 + per-agent T015/T017(6)/T020(6) |
| FR-016 sentinel | Symlinked `SKILL.md` fails lstat check | `claude.test.ts` T015 sentinel row |
| FR-016 inner | Interior symlinks filtered from tar | T014(6) + mirrors + `tar.test.ts` T004 |
| FR-017 | Top-level dot entries silently skipped | walker + Codex `.system` T018 |
| SC-009 | Vault never contains vendored-pool content via symlinked root | `integration.test.ts` T026 SC-009 test |

## Implementation notes

- **Walker is the single source of truth.** Every `snapshot<Agent>` call now delegates skill collection to `collectSkillArtifacts("<agent>", AgentPaths.<agent>.skillsDir)`. A future CVE in any gate is patched in one file instead of four.
- **`tar.ts::archiveDirectory` filter is opt-in.** The existing Copilot agent-tarball callers (pre-walker) keep their current bit-for-bit behavior — the `skipSymlinks` flag is only passed inside the walker's skill archival path.
- **`performSkillRemove` returns a discriminated result.** Every contract row is a deterministic branch (`success`, `unknown-agent`, `not-found`, `reconcile-error`, `git-error`). The citty wrapper translates branches to `log.*` + `process.exitCode`.
- **`readHeadShortSha` is non-fatal.** If git fails to read the post-commit short SHA for the success log line, the command still reports success without the SHA fragment — the SHA is cosmetic, not a correctness signal.
- **Two Mermaid diagrams in `docs/architecture.md`** (sync flow and vault-removal flow) match the project's GitHub-compatible Mermaid rules: `<br/>` line breaks, inline `:::className` styles, single `subgraph` per diagram, `classDef` pairs meeting WCAG 2 AA contrast, descriptive node IDs ≥ 3 chars.

## Test plan

- [x] `bun run check` green — 393 pass / 0 fail / 850 expect calls / 39 files
- [x] Coverage floors met: `skills-walker.ts` 90.70%, `skill.ts` 78.68%, `claude.ts` 90.43%, `codex.ts` 88.00%, `cursor.ts` 95.21% (all ≥ 70% Principle II), `sanitizer.ts` + `encryptor.ts` still 100% (≥ 90% security-critical)
- [x] Two Mermaid diagrams reviewed against the existing reconciliation flowchart's known-good grammar (same `flowchart TD` / inline `:::className` / `classDef` patterns). *Note: the MCP Mermaid validator returned a Puppeteer/Chromium infrastructure error (`spawn /usr/bin/chromium EAGAIN`), not a syntax error — the diagrams should be re-verified in GitHub's markdown preview as part of PR review.*
- [ ] **T034 — manual cross-machine quickstart walkthrough**: reviewer or author to run every step in `specs/20260411-002222-agent-skills-sync/quickstart.md` end-to-end against two isolated `\$HOME` sessions (or two physical machines) and paste the resulting pass/fail matrix into this PR description.
- [ ] Verify `agentsync skill --help` and `agentsync skill remove --help` both render after the CLI registration change in `src/cli.ts`.
- [ ] Smoke-test `skill remove` against a real vault with one skill artifact, confirm the commit lands and the local skill directory survives.

## Safety notes for reviewers

- No destructive git operations are executed automatically. `skill remove` is the only new command that writes to the vault, and it is strictly one-skill-at-a-time with no `--all` / `--force` / glob support.
- The pre-commit hooks (biome + typecheck + conventional-commit message) ran green on this commit.
- `.claudeignore` was intentionally NOT staged — it is a pre-existing untracked file unrelated to this feature.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Sync user-created "skills" for Claude, Codex, Cursor (and retrofitted Copilot) into the encrypted vault
  * New CLI: agentsync skill remove <agent> <name> to delete vault-only skill artifacts

* **Behavior Changes**
  * Push aborts if forbidden/never-sync files are found inside a skill
  * Pull is extract-only — vault removals never delete local skill directories
  * Archives omit symlinked helper files; skill names validated to prevent path traversal
  * Canonical vault layout: <vaultDir>/<agent>/skills/<name>.tar.age

* **Documentation**
  * Added command reference, architecture diagrams, troubleshooting, quickstart, README updates

* **Doctor / Status**
  * Doctor checks per-agent skills directories; Status reports skill drift alongside other vault entries

* **Tests**
  * Extensive unit/integration tests covering snapshot, apply, walker, CLI remove, push, pull, and edge cases

### CODE REVIEW

File: docs/command-reference.md [hunk 1]
```diff
@@ -94,14 +97,15 @@ bunx --package @chrisleekr/agentsync agentsync pull --agent cursor
...
-**Outcome**: supported local agent files are updated from the vault.
+**Outcome**: supported local agent files are updated from the vault, including any per-agent skills stored under `claude/skills/`, `codex/skills/`, `cursor/skills/`, and `copilot/skills/`.
 
 **Caveats**:
 
 - Pull applies only enabled or explicitly requested agents.
 - Pull uses the same fast-forward-only reconciliation rule as `push`, `key`, and daemon sync.
 - If local and remote vault history diverged, `pull` exits with a recovery message and no success footer.
 - If the private key is missing, the command cannot decrypt anything.
+- Pull is extract-only and never deletes an existing local skill directory, even when the matching vault artifact is gone (FR-013). The only way to remove a skill everywhere is the explicit `skill remove` verb.
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply docs/command-reference.md:109]: Fixed in c4d1d53. The `pull` caveat no longer claims `skill remove` removes a skill "everywhere" — it now clearly states that `skill remove` only removes the vault artifact and that local directories must be deleted by hand with `rm -rf ~/.<agent>/skills/<name>` on each machine. Thanks for catching the contradiction with FR-013.

File: docs/command-reference.md [hunk 2]
```diff
@@ -133,8 +137,41 @@ bunx --package @chrisleekr/agentsync agentsync doctor
...
+**Outcome on success**: the file `<vaultDir>/<agent>/skills/<name>.tar.age` is removed, a `skill remove(<agent>): <name>` commit is created, and the commit is pushed to the configured remote branch. The success log line includes the 7-character short SHA.
+
+**Exit codes**:
+
+| Code | Scenario | Output |
+| ---- | -------- | ------ |
+| `0` | File removed, commit landed, push succeeded | `Removed <agent>/<name> from vault (commit <sha7>)` |
+| `1` | Skill not found in vault | `Skill not found: <agent>/<name>` + `Looked for: <resolved path>` |
+| `1` | Unknown agent name | `Unknown agent: <provided>. Supported: claude, cursor, codex, copilot` |
+| `1` | Reconcile or git push failed | `Removal staged but not pushed: <git error>` with retry hint |
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply docs/command-reference.md]: Fixed in c4d1d53. The exit-code table in `docs/command-reference.md` now has two distinct rows: one for `reconcile-error` (no staged change, raw error message) and one for `git-error` (staged deletion, "Removal staged but not pushed:" prefix + retry hint). The wording mirrors what `src/commands/skill.ts:184-196` actually logs and matches the authoritative `contracts/skill-remove-cli.md:35-36` contract. Added a pointer to the contract below the table so future readers land on the source of truth.

File: specs/20260411-002222-agent-skills-sync/contracts/walker-interface.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+ *                     returns { artifacts: [], warnings: [] }.
+ *
+ * @returns            SkillsWalkerResult with one artifact per qualifying
+ *                     skill and zero or more `never-sync inside skill: `
+ *                     warnings for skills rejected at gate 4.
+ */
+export async function collectSkillArtifacts(
+  agent: AgentName,
+  skillsDir: string,
+): Promise<SkillsWalkerResult>;
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply specs/20260411-002222-agent-skills-sync/contracts/walker-interface.md]: Fixed in c4d1d53. The contract now documents `agent: SkillBearingAgent` to match `src/agents/skills-walker.ts:71-72`. Dropped the "callers that pass vscode get an empty result" language because the type-level exclusion makes that runtime no-op unreachable. Thanks — good catch on the spec/impl drift.

File: specs/20260411-002222-agent-skills-sync/data-model.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+/**
+ * Result shape for the `skill remove <agent> <name>` command.
+ * Modelled as a tagged union so the CLI layer can decide exit code +
+ * output format without re-parsing a message string.
+ */
+type RemovalOutcome =
+  | { kind: "removed"; agent: AgentName; name: string; vaultPath: string; commitSha: string }
+  | { kind: "not-found"; agent: AgentName; name: string; vaultPath: string }
+  | { kind: "git-error"; agent: AgentName; name: string; vaultPath: string; error: string };
+```
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply specs/20260411-002222-agent-skills-sync/data-model.md]: Fixed in c4d1d53. `data-model.md` now exports `SkillRemoveResult` with a `status` discriminant and the full 5-variant union (`success` / `unknown-agent` / `not-found` / `reconcile-error` / `git-error`), matching `src/commands/skill.ts:21-26` exactly. Also added an inline "spec note" explaining the history of the rename so the next reader can see why the earlier draft had a narrower 3-variant `kind` union, and a cross-reference to `contracts/skill-remove-cli.md` as the canonical CLI-mapping doc.

File: src/agents/__tests__/skills-walker.test.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    await mkdir(skillDir, { recursive: true });
+    await writeFile(join(skillDir, "SKILL.md"), "# dirty", "utf8");
+    await writeFile(join(skillDir, "auth.json"), '{"token":"x"}', "utf8");
+
+    const result = await collectSkillArtifacts("claude", tmpDir);
+    expect(result.artifacts).toHaveLength(0);
+    expect(result.warnings.length).toBeGreaterThanOrEqual(1);
+    const offending = result.warnings.find((w) => w.startsWith("never-sync inside skill: "));
+    expect(offending).toBeDefined();
+    expect(offending).toContain("auth.json");
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/agents/__tests__/skills-walker.test.ts:179]: Fixed in c4d1d53. Rows 11 and 12 in `src/agents/__tests__/skills-walker.test.ts` now use `toHaveLength(1)` and assert the warning prefix on the indexed element directly, matching the walker-interface contract's "exactly one warning per offending path" guarantee. A future regression that duplicates warnings will now be caught.

File: src/agents/codex.ts [hunk 6]
```diff
@@ -123,6 +134,24 @@ export async function applyCodexRule(ruleName: string, content: string): Promise
...
+ * @param skillName  Basename of the skill (no extension).
+ * @param base64Tar  Base64-encoded `.tar.gz` payload that the walker
+ *                   produced on the source machine.
+ */
+export async function applyCodexSkill(skillName: string, base64Tar: string): Promise<void> {
+  const targetDir = join(AgentPaths.codex.skillsDir, skillName);
+  await mkdir(targetDir, { recursive: true });
+  const tarBuffer = Buffer.from(base64Tar, "base64");
+  await extractArchive(tarBuffer, targetDir);
+}
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/agents/codex.ts:154]: Fixed in c4d1d53 — and the scope is wider than originally flagged: the same `basename(name, ".tar.age")` → `join(skillsRoot, ..)` pattern exists identically in `applyClaudeSkill`, `applyCursorSkill`, and `applyCopilotSkill`, so all four were vulnerable. Rather than adding an inline check in each, I added a shared `validateSkillName` helper + `InvalidSkillNameError` class in `src/agents/skills-walker.ts` (same module that already owns the push-side gates), and wired it into every `applyXxxSkill` function plus the corresponding skill loop in each `applyXxxVault`. Rejects empty, `.`, `..`, leading-dot, path separators, and control characters.

Walked the exact exploit: a vault file literally named `...tar.age` → `basename` strips the 8-char suffix → `skillName === ".."` → `join(skillsRoot, "..")` = agent config root → `extractArchive` would have written tar entries like `AGENTS.md` or `config.toml` on top of legitimate user config. The existing tar-slip filter in `src/core/tar.ts:92-98` defends against malicious entry paths inside the tar, but does NOT sanitize the `cwd` argument — so the validator is the right place to close the gap.

Added adversarial-filename regression tests for all 4 agents that seed a vault with a `...tar.age` file containing a payload that would overwrite each agent's config root, then assert (a) `applyXxxSkill("..", ...)` throws `InvalidSkillNameError` and (b) `applyXxxVault` logs a skip warning and leaves the parent of `skillsDir` untouched. See `src/agents/__tests__/{claude,codex,cursor,copilot}.test.ts`. Tests: 402 pass / 0 fail / 889 expect calls (was 393/0/850).

File: src/agents/cursor.ts [hunk 7]
```diff
@@ -117,6 +129,25 @@ export async function applyCursorCommand(commandName: string, content: string):
...
+ * @param skillName  Basename of the skill (no extension).
+ * @param base64Tar  Base64-encoded `.tar.gz` payload that the walker
+ *                   produced on the source machine.
+ */
+export async function applyCursorSkill(skillName: string, base64Tar: string): Promise<void> {
+  const targetDir = join(AgentPaths.cursor.skillsDir, skillName);
+  await mkdir(targetDir, { recursive: true });
+  const tarBuffer = Buffer.from(base64Tar, "base64");
+  await extractArchive(tarBuffer, targetDir);
+}
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/agents/cursor.ts:150]: Thanks for the catch — this is the one review item from this pass I deliberately did **not** fix in c4d1d53 because the proposed `rm -rf` + extract semantic change conflicts with the spec's additive-default posture, and I think it needs a design decision rather than a silent code change. Leaving this thread **unresolved** for your direction.

**Observation you flagged** (real): extracting the vault tar on top of an existing local skill directory leaves stale files — files that existed locally but aren't in the tar remain. Same pattern in `applyClaudeSkill`, `applyCodexSkill`, `applyCopilotSkill` (not just cursor).

**Why I don't think `rm -rf` + extract is a drop-in fix**: FR-011 ("additive by default") and FR-013 ("pull is extract-only, never deletes local") together imply the pull path is supposed to preserve local edits rather than enforce vault-as-source-of-truth. Concretely, if a user edited `~/.cursor/skills/my-skill/helper.md` locally before the next pull, `rm -rf` would destroy that work silently. The current overlay behavior trades consistency (stale files survive when upstream deletes) for safety (local edits are never lost).

**Possible resolutions, in order of disruption**:
1. **Keep current behavior, document the drift.** Add a troubleshooting entry explaining that a skill deletion-of-files on machine A won't propagate to machine B via `pull` — the user must `rm -rf` the stale file locally (symmetric to the skill-level guarantee already documented for FR-013).
2. **Atomic replace via tmp-dir rename.** Extract into a sibling tmp dir, then `rename` on top of the target. Still destroys local edits, but at least there's no half-extracted state on failure. Same FR-011 tension.
3. **Per-file diff + opt-in delete.** Extract-then-reconcile, keeping local-only files that don't appear in the tar. Closest to the FR-011 spirit but requires tracking tar-entry-set and is the biggest behavior change.
4. **User-facing flag.** `pull --replace-skills` for users who want vault-as-source-of-truth; default stays additive.

Happy to go in any direction you pick — just not willing to quietly change a stable feature's semantics. If option 1 is enough, I can land the troubleshooting entry in this PR.

File: src/commands/doctor.ts [hunk 8]
```diff
@@ -11,12 +11,50 @@ import { AgentPaths } from "../config/paths";
...
+  for (const [name, dir] of targets) {
+    try {
+      await access(dir, constants.R_OK);
+      checks.push({ name, status: "pass", detail: dir });
+    } catch {
+      checks.push({
+        name,
+        status: "warn",
+        detail: `Not found or unreadable: ${dir}`,
+      });
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/commands/doctor.ts:77]: Fixed in c4d1d53. `buildSkillsDirChecks` in `src/commands/doctor.ts` now calls `stat` after `access` and pushes a `warn` row with detail "Exists but is not a directory" when the target is a regular file instead of a directory. Added a new test (`reports warn when a skills path exists but is not a directory`) that seeds a regular file at the claude skills path and asserts the warn outcome.

File: src/agents/copilot.ts [hunk 9]
```diff
@@ -234,9 +212,18 @@ export async function applyCopilotVault(
...
+    const skillName = basename(name, ".tar.age");
+    try {
+      validateSkillName(skillName);
+    } catch (err) {
+      if (err instanceof InvalidSkillNameError) {
+        log.warn(`[copilot] Skipping vault skill with invalid name '${name}': ${err.reason}`);
+        continue;
+      }
+      throw err;
+    }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/agents/copilot.ts:230]: Fixed in e97d456. You were right — the `agents/` loop in `applyCopilotVault` had the exact same `basename(name, ".tar.age")` → `path.join(agentsDir, agentName)` path-traversal surface as the `skills/` loop, and I missed it in the Phase 8 rollout. A vault file literally named `...tar.age` collapsed to `..`, escaped `agentsDir`, and `extractArchive` would have written tar entries into the Copilot config root (e.g. overwriting `~/.copilot/instructions`).

The fix wires `validateSkillName` into both `applyCopilotAgent`'s entry point and the `agents/` loop in `applyCopilotVault`, symmetric with the already-fixed `skills/` side. Two regression tests were added to `copilot.test.ts`:

1. A direct-function assertion using the same `badNames` array the skills loop uses (`["", ".", "..", "../foo", "foo/bar", "foo\\bar", ".hidden", "foo\x00bar"]`) that confirms `applyCopilotAgent` throws `InvalidSkillNameError` on each one.
2. An end-to-end adversarial-filename test that writes `...tar.age` into `<vault>/copilot/agents/` with a distinct `AGENT_LEAKED.md` payload and asserts no file escapes to the parent of `agentsDir`.

Note on naming: `validateSkillName` is now slightly stretched semantically — it validates both skill and agent basenames. The function is really a generic vault-basename check and could be renamed in a follow-up, but I kept the existing export to avoid touching the 4 other call sites that already rely on it.

Thanks for catching this — a HIGH-severity path traversal that shipped quietly alongside the skills fix would have been genuinely bad.

File: src/commands/doctor.ts [hunk 10]
```diff
@@ -11,12 +11,61 @@ import { AgentPaths } from "../config/paths";
...
+      const info = await stat(dir);
+      if (!info.isDirectory()) {
+        checks.push({
+          name,
+          status: "warn",
+          detail: `Exists but is not a directory: ${dir}`,
+        });
+        continue;
+      }
+      checks.push({ name, status: "pass", detail: dir });
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/commands/doctor.ts:71]: Fixed in 596581e. You were right — `doctor` was using `stat()` which follows the symlink and reports a real directory, so a user with `~/.claude/skills -> /srv/team-pool` would see a `pass` row and then have `push` silently sync nothing because the walker's `lstat` check at `skills-walker.ts:150-151` rejects symlinked roots per FR-016.

Switched to `lstat()` and added a dedicated `warn` branch with detail `Symlinked skills root is not synced (FR-016): <dir>` so the user can trace the behaviour directly back to the spec rule. Regression test creates a real target directory, symlinks `skillsDir` to it, and asserts the warn row has both "Symlinked skills root" and "FR-016" in the detail — the FR reference gives the user a searchable term if they want to understand why the walker skips the path.

Thanks for the catch — inconsistency between doctor and walker on the same path would have been a confusing user experience.

File: src/agents/cursor.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/agents/cursor.ts:150]: Option 1 landed as cfd0e83. The troubleshooting entry covers exactly the symptom/cause/resolution you suggested:

- **Symptom**: file deleted from a skill on machine A, vault updated via push, but the file still present on machine B after pull.
- **Cause**: `applyXxxSkill` extracts the tar on top of the existing local directory; files that exist locally but are no longer in the tar survive the extract (FR-013 extract-only guarantee).
- **Resolution**: manually `rm` the stale file on the affected machine, then `agentsync status` to confirm.

I placed the new section between the two existing FR-013 entries (`I deleted a skill locally and it came back after pull` and `I removed a skill from the vault but it is still on my other laptop`) so readers find all three extract-only drift scenarios grouped together. I also flagged that a future `pull --replace-skills` flag could revisit this once real-world usage tells us whether the drift is common enough to justify the added complexity, but the default stays additive.

Thanks for the back-and-forth on this one — cross-referencing FR-011/FR-013 before landing a silent behaviour change was the right call, and the troubleshooting note is a much cheaper fix than the 4 options I listed.

File: docs/command-reference.md [hunk 11]
```diff
@@ -133,8 +137,44 @@ bunx --package @chrisleekr/agentsync agentsync doctor
 - age-encryption module availability
 - remote reachability
 - obvious unencrypted sensitive files in the vault
+- readability of the per-agent skills directories under `~/.claude/skills/`, `~/.codex/skills/`, and `~/.cursor/skills/` (warns if a directory is missing or unreadable)
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply docs/command-reference.md]: Fixed in bc0c8f9. Bullet now names `buildSkillsDirChecks` and lists all three warn cases (missing/unreadable, not-a-directory, symlinked root per FR-016). Thanks for catching the drift between the doctor rule and this bullet.

File: docs/command-reference.md [hunk 12]
```diff
@@ -133,8 +137,44 @@ bunx --package @chrisleekr/agentsync agentsync doctor
...
+```bash
+bunx --package @chrisleekr/agentsync agentsync skill remove claude my-skill
+bunx --package @chrisleekr/agentsync agentsync skill remove codex review-helper
+bunx --package @chrisleekr/agentsync agentsync skill remove cursor polish-prompt
+bunx --package @chrisleekr/agentsync agentsync skill remove copilot debug-flow
+```
+
+**Signature**: `skill remove <agent> <name>` — both positional arguments are required. `<agent>` must be one of `claude`, `cursor`, `codex`, or `copilot`; `vscode` and any other value are rejected. `<name>` is the basename of the skill (no extension, no path separators).
+
+**Outcome on success**: the file `<vaultDir>/<agent>/skills/<name>.tar.age` is removed, a `skill remove(<agent>): <name>` commit is created, and the commit is pushed to the configured remote branch. The success log line includes the 7-character short SHA.
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply docs/command-reference.md]: Fixed in 2faa480. Wording now reflects that `readHeadShortSha` is intentionally non-fatal — the `(commit <sha7>)` fragment only appears when `result.commitSha` is truthy (see `src/commands/skill.ts:167`). Thanks for catching the overstated guarantee.

File: specs/20260411-002222-agent-skills-sync/contracts/walker-interface.md [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+| 3 | Real skill `my-skill/SKILL.md` + `my-skill/README.md` | 1 | `[]` |
+| 4 | Real skill `my-skill/` with no `SKILL.md` | 0 | `[]` |
+| 5 | Real skill `my-skill/` where `SKILL.md` is a symlink to another file | 0 | `[]` |
+| 6 | Top-level symlink `my-skill -> /tmp/other/my-skill` with valid target | 0 | `[]` |
+| 7 | Top-level `.system/` directory containing a real skill | 0 | `[]` |
+| 8 | Top-level `.DS_Store` regular file | 0 | `[]` |
+| 9 | Real skill `a/SKILL.md` + another real skill `b/SKILL.md` + symlinked root `c/` | 2 | `[]` |
+| 10 | Real skill whose interior contains a symlinked helper file | 1 | `[]` (the helper is omitted silently) |
+| 11 | Real skill containing a file whose path matches `NEVER_SYNC_PATTERNS` (e.g. `./credentials.json`) | 0 (the skill is not archived) | 1 warning with prefix `never-sync inside skill: ` |
+| 12 | Two real skills, one of which has a never-sync match | 0 for the matched skill, 1 for the clean one → `artifacts.length === 1` | 1 warning with prefix `never-sync inside skill: ` |
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply specs/20260411-002222-agent-skills-sync/contracts/walker-interface.md]: Fixed in 5d1b103. Replaced the trailing-space code spans with the full emitted warning shape `never-sync inside skill: <absolute path>` — this satisfies MD038 and also documents the complete warning format from `src/agents/skills-walker.ts:200` instead of just the prefix fragment.

File: specs/20260411-002222-agent-skills-sync/data-model.md [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+  artifacts: SkillArtifact[];
+  warnings: string[];
+}
+```
+
+### Invariants
+
+- `artifacts` contains one entry per **Skill on disk that passed all walker gates**. Symlinked roots, dot-prefixed entries, directories without a real `SKILL.md`, and directories whose interior walk found a never-sync match are all **absent** from `artifacts`.
+- `warnings` contains zero or more strings. Walker-produced warnings use the prefix `"never-sync inside skill: "` followed by the absolute path that matched and the pattern it matched. No other prefixes are produced by the walker — dot-skip and symlink-skip are silent (FR-017, R7).
+- `artifacts` and `warnings` are independent: a single push can produce both (for example, three valid skills get archived AND one skill is rejected for a never-sync hit; the push will still abort before encrypting the three valid skills because the push gate escalates any walker warning with the never-sync prefix to fatal — R3).
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply specs/20260411-002222-agent-skills-sync/data-model.md:97]: Fixed in 81b90e6. Struck "and the pattern it matched" — `src/agents/skills-walker.ts:200` emits only `${NEVER_SYNC_WARNING_PREFIX}${hit}` where `hit` is just the absolute path from `collectNeverSyncHits`. The pattern never leaves the walker. Thanks for catching this spec drift.

File: src/agents/skills-walker.ts [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+  for (let i = 0; i < name.length; i++) {
+    const code = name.charCodeAt(i);
+    if (code < 0x20) {
+      throw new InvalidSkillNameError(name, "contains control character");
+    }
+    if (code === 0x2f || code === 0x5c) {
+      throw new InvalidSkillNameError(name, "contains path separator");
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/agents/skills-walker.ts:111]: Declining this expansion for now — no Windows code change in this PR. Two reasons:

1. **No Windows test target available.** We cannot verify reserved-char / reserved-device-name handling against real Windows FAT32/NTFS semantics, so adding the rules blind would give a false sense of coverage.

2. **The feature is currently POSIX-shaped.** `AgentPaths` hard-codes paths like `~/.claude/skills/...`, `~/.cursor/skills/...`, `~/.codex/skills/...`, `~/.copilot/skills/...` (see `src/config/paths.ts`). The skills sync feature has never been validated on Windows as a whole, so tightening `validateSkillName` for Windows-reserved names specifically would be a half-measure against an unsupported target.

If Windows becomes a supported target in a future release, the validator expansion is well-scoped and easy to add as a follow-up:
- Deny the Windows reserved characters: `:`, `*`, `?`, `"`, `<`, `>`, `|`
- Deny trailing dot or trailing space (stripped by Windows at creation time)
- Deny the reserved device names `CON`, `PRN`, `AUX`, `NUL`, `COM1..COM9`, `LPT1..LPT9` (case-insensitive, with or without extension)

For now, `validateSkillName` stays POSIX-shaped to match the rest of the feature. Thanks for the thorough review — the pointer is preserved in this thread for whoever picks Windows support up later.
